plugin/finder.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+var validPluginFilenamePrefixes = []string{"trellis"}
+
+type finder struct {
+	searchPaths []string
+	verifier    pathVerifier
+}
+
+func newFinder(paths []string) finder {
+	return finder{
+		searchPaths: paths,
+		verifier:    &binPathVerifier{},
+	}
+}
+
+func (o *finder) find() map[string]string {
+	plugins := make(map[string]string)
+
+	for _, dir := range uniquePathsList(o.searchPaths) {
+		if len(strings.TrimSpace(dir)) == 0 {
+			continue
+		}
+
+		files, err := ioutil.ReadDir(dir)
+		if err != nil {
+			continue
+		}
+
+		for _, f := range files {
+			if f.IsDir() {
+				continue
+			}
+			if !hasValidPrefix(f.Name(), validPluginFilenamePrefixes) {
+				continue
+			}
+
+			// TODO: Handle overshadowed plugins, i.e: plugins with the same binary name.
+			if ok, name, path := o.verifier.verify(filepath.Join(dir, f.Name())); ok {
+				plugins[name] = path
+			}
+		}
+	}
+
+	return plugins
+}
+
+// pathVerifier receives a path and determines if it is valid or not
+type pathVerifier interface {
+	// verify determines if a given path is valid
+	verify(path string) (bool, string, string)
+}
+
+type binPathVerifier struct{}
+
+// verify implements pathVerifier and determines if a given path
+// is valid depending on whether or not it overwrites an existing
+// trellis command path, or a previously seen plugin.
+func (v *binPathVerifier) verify(path string) (bool, string, string) {
+	// extract the plugin binary name
+	segments := strings.Split(path, "/")
+	binName := segments[len(segments)-1]
```

Comment:
`filepath.Base(path)` instead? It handles Windows path separators too.

Reply:
Good catch